docs/DocumenterReference.jl

```diff
@@ -0,0 +1,198 @@
+
+Documenter.Selectors.order(::Type{APIBuilder}) = 0.0
+
+"""
+    automatic_reference_documentation(
+        current_module::Module;
+        root::String,
+        subdirectory::String,
+        extras::Vector{Pair{String,DocType}} = Pair{String,DocType}[],
+    )
+
+Automatically creates the API reference documentation for `current_module` and
+returns a `Vector` which can be used in the `pages` argument of
+`Documenter.makedocs`.
+
+## Arguments
+
+ * `current_module`: the module from which to create an API reference.
+ * `subdirectory`: the directory relative to the documentation root in which to
+   write the API files.
+ * `extras`: a vector of non-exported docstrings to include in the API
+   reference. Each element is a pair which maps the docstring signature to a
+   [`DocumenterReference.DocType`](@ref) enum.
+
+## Multiple instances
+
+Each time you call this function, a new object is added to the global variable
+`DocumenterReference.CONFIG`.
+"""
+function automatic_reference_documentation(
+    current_module::Module;
+    root::String,
+    subdirectory::String,
+    extras::Vector{Pair{String,DocType}} = Pair{String,DocType}[],
+)
+    config = _Config(current_module, root, subdirectory, extras)
+    pages = Any["Overview"=>"$subdirectory/$current_module.md"]
+    _iterate_over_symbols(config) do key, type
+        push!(pages, Documenter.hide("`$key`" => "$subdirectory/$key.md"))
+        return
+    end
+    push!(CONFIG, config)
+    return pages
+end
+
+function _exported_symbols(mod)
+    contents = Pair{Symbol,DocType}[]
+    for n in names(mod)
+        f = getfield(mod, n)
+        f_str = string(f)
+        if startswith(f_str, "@")
+            push!(contents, n => DOCTYPE_MACRO)
+        elseif startswith(f_str, "Abstract")
+            push!(contents, n => DOCTYPE_ABSTRACT_TYPE)
+        elseif f isa Type
+            push!(contents, n => DOCTYPE_STRUCT)
+        elseif f isa Function
+            if islowercase(f_str[1])
+                push!(contents, n => DOCTYPE_FUNCTION)
+            else
+                push!(contents, n => DOCTYPE_STRUCT)
+            end
+        elseif f isa Module
+            # Skip
+        else
+            push!(contents, n => DOCTYPE_CONSTANT)
+        end
+    end
+    order = Dict(
+        DOCTYPE_MACRO => 1,
+        DOCTYPE_FUNCTION => 2,
+        DOCTYPE_ABSTRACT_TYPE => 3,
+        DOCTYPE_STRUCT => 4,
+        DOCTYPE_CONSTANT => 5,
+    )
+    return sort(contents; by = x -> (order[x[2]], "$(x[1])"))
+end
+
+function _iterate_over_symbols(f, config)
+    current_module = config.current_module
+    for (key, type) in vcat(_exported_symbols(current_module), config.extras)
+        if key isa Symbol
+            doc = Base.Docs.doc(Base.Docs.Binding(current_module, key))
+            if occursin("No documentation found.", string(doc))
+                if type == DOCTYPE_CONSTANT
+                    continue  # It's okay not to document every constant.
```

Comment:
I'd like to remove this. Everything in the public API should be documented. But it requires https://github.com/jump-dev/MathOptInterface.jl/pull/2186 first.